### PR TITLE
Fix changing column order/visibility via sidebar is inconsistent #13455

### LINKED ISSUE #13455: Changing column order/visibility via visualization Settings sidebar is inconsistent
**Describe the bug**
When using the visualization Settings sidebar to reorder column and removing columns, causes some inconsistency. Reordering works, but removing columns will reset the order. And every change will show Play-button instead of just doing the change.

**To Reproduce**
1. Custom question > Sample Dataset > Orders - join table Products
image
2. Click visualization Settings sidebar > drag "Product -> Category" column to beginning of table - it actually moves the column, when looking underneath the overlay, but still shows Play-button (the old "Get answers"):
image
3. Click the Play-button and now remove the column "Subtotal", which will then seem to do it correctly underneath the overlay:
image
4. But after clicking the Play-button, then "Subtotal" is removed, but any column ordering has reverted to default:
image

**Information about your Metabase Installation:**
Metabase 0.35.4, 0.36.7, 0.37.0-rc2

**Additional context**
Related to #12395. And a little to #12044.
This is not a regression, but it's basically some of the same problems as #10466

### PR DESCRIPTION
**Pull request by WiNloSt**
Fixes #13455

I'm not exactly sure if this will break anything else. But basically, there are 3 problems in #13455
1. Changing column order the first time will always show the "Play-button"
1. Changing joined table column order after the first time will show the "Play-button".
    
1. Changing column order and visibility will revert the order after clicking the "Play-button".

So, there are 2 approaches to fixing this issue.
1. Make it so that the "Play-button" never shows.
2. Fix the result after clicking "Play-button" given there are column order and visibility settings changed to be consistent.

Let's see what causes each problem in detail.

## 1. Changing column order the first time will always show the "Play-button"
The root cause is from how we're determining if the visualization is *dirty* or not.
https://github.com/metabase/metabase/blob/e6c0cbfdd44460d7f98bbfd40e4fbe1b11bc12d1/frontend/src/metabase/query_builder/selectors.js#L361-L366

The first time when we change the column order the `query.joins[int].fields` will change from `all` to an array of fields
### Before column reordering
image

### After column reordering
image

There's no way at the moment to verify if the `all` and an array of fields are equal. Or it would be hacky.

## 2. Changing joined table column order after the first time will show the "Play-button".
I traced it back to https://github.com/metabase/metabase/issues/10466 which [added the logic to sort `query.fields`](https://github.com/metabase/metabase/pull/11311/files#diff-2fc40467e8370b516c4ce51320cd4d159278808493231640527a65f51d06e65aR170). But there's no place that we sort `query.joins[int].fields`. That's why the "Play-button" will only appear when we sort the joined table columns, but not the primary table columns.

## 3. Changing column order and visibility will revert the order after clicking the "Play-button".
It's because of how we compute the series settings.
From here
https://github.com/metabase/metabase/blob/8220d653e7fde7eb1797a9a515b4fda5126ed76f/frontend/src/metabase/visualizations/components/Visualization.jsx#L147
it eventually calls
https://github.com/metabase/metabase/blob/bceca3b31e90e6ab5d23ed4937b439e466baefbd/frontend/src/metabase/visualizations/lib/settings.js#L94-L98
which calls
https://github.com/metabase/metabase/blob/444d16c81b7d31c83e1bdc9d3aa926097b2769c7/frontend/src/metabase/visualizations/visualizations/Table.jsx#L169-L179

if you look carefully at the validator
https://github.com/metabase/metabase/blob/444d16c81b7d31c83e1bdc9d3aa926097b2769c7/frontend/src/metabase/visualizations/visualizations/Table.jsx#L175-L179
it tries to match column settings with data columns which are changed after we run the query after hiding a column.

image

You could see that this will always fail as we have less `data.cols` than the column settings. I'm not sure if this will break other functionality or not, but it seems to me we should only test the enabled column. As those columns with `enabled: false` will not be present in the `data.cols`
image

## Summary
This PR addresses 2. and 3. but not 1. as it would require the work to match `"all"` with an array of fields.

### CODE REVIEW

File: frontend/src/metabase/query_builder/selectors.js [hunk 1]
```diff
@@ -314,6 +315,23 @@ export function normalizeQuery(query, tableMetadata) {
         );
       });
     }
+
+    // sort query.join[int].fields
+    if (query.query.joins) {
+      query = updateIn(query, ["query", "joins"], joins =>
+        joins.map(joinedTable => {
+          if (!joinedTable.fields || joinedTable.fields === "all") {
```
Comment (WiNloSt) [inline frontend/src/metabase/query_builder/selectors.js:323]: 1. The first predicate is to prevent [this test](https://github.com/metabase/metabase/runs/5751459337?check_suite_focus=true) from failing
2. I found the joined table's `fields` could be a string `"all"`

image

**Comment by WiNloSt** [comment]
[`questions/settings/cy.spec.js` seems to flake](https://app.circleci.com/pipelines/github/metabase/metabase/32130/workflows/db05248f-0cb7-480b-9591-770e9f871e4a/jobs/1639793).

File: frontend/test/metabase/scenarios/question/settings.cy.spec.js [hunk 2]
```diff
@@ -138,6 +138,10 @@ describe("scenarios > question > settings", () => {
         cy.icon("play")
           .last()
           .click();
+
+        // Prevent performing actions while the query is being executed.
+        // Which caused some race condition and failed the test.
+        cy.findByText(/Doing science/).should("not.exist");
```
Comment (WiNloSt) [inline frontend/test/metabase/scenarios/question/settings.cy.spec.js:144]: This is the culprit for [all](https://app.circleci.com/pipelines/github/metabase/metabase/32363/workflows/2df69eb8-cd75-4ffc-a3e3-5dc276f64d59/jobs/1654770) [the](https://app.circleci.com/pipelines/github/metabase/metabase/32363/workflows/2df69eb8-cd75-4ffc-a3e3-5dc276f64d59/jobs/1654805) [failed](https://app.circleci.com/pipelines/github/metabase/metabase/32130/workflows/db05248f-0cb7-480b-9591-770e9f871e4a/jobs/1639793) [tests](https://app.circleci.com/pipelines/github/metabase/metabase/32122/workflows/ad5b2fff-02af-43d9-86f9-ac80f4934535/jobs/1639214).

Basically, it's a race condition where Cypress removes the field before the query is finished executing, causing the `card` to be overridden by an older version (before removing the field).

**Review by flamber — COMMENTED** [review]
While it fixes some things it also causes newly added column to be unmovable.
I think you're close to solve it.

https://user-images.githubusercontent.com/1447303/161256525-31a8a4cf-df23-4fb9-804b-6e67557a80aa.mp4

File: frontend/test/metabase/scenarios/question/settings.cy.spec.js (hunk 2 above)
Comment (flamber) [reply frontend/test/metabase/scenarios/question/settings.cy.spec.js:144]: @nemanjaglumac Seems like tests in settings.cy.spec.js could use a little race-condition tweaking by waiting for requests.
Also note the fail on Electron and not on Chrome, which we talked about earlier this week.

**Comment by WiNloSt** [comment]
> While it fixes some things it also causes newly added column to be unmovable. I think you're close to solve it.
> 
>  Screencast-2022-04-01_13.39.41.mp4

Well... I'll give it a look next week. Thanks for finding this out.

File: frontend/test/metabase/scenarios/question/settings.cy.spec.js (hunk 2 above)
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/question/settings.cy.spec.js:144]: This will do the trick and is more reliable than asserting on the "non-existence" of a UI element.
Btw, if you're wondering where does that intercept come from - [`visitQuestionAdHoc()`](https://github.com/metabase/metabase/blob/aa0878e372e2a1458ab0e2f2bfe592c45094e280/frontend/test/__support__/e2e/helpers/e2e-ad-hoc-question-helpers.js#L29) helper sets the "spy" / "listener" on `POST /api/dataset` route and we can intercept it later in the test "for free".

```suggestion

        cy.wait("@dataset");
```

Comment (WiNloSt) [reply frontend/test/metabase/scenarios/question/settings.cy.spec.js:144]: Thanks, I never knew about that intercept. Although, I wonder what do you mean by using spy is more *reliable* than asserting the "non-existence" of a UI element. As for me, I kind of felt like using spy is some kind of smell, because Cypress should represent what users see, so only the UI makes sense. Spying the APIs means we're leaking the implementation details to tests, which should be used as a last resort.

For me spy would make more sense when you want to:
- Simulate race condition
- Other complex states due to asynchronous actions

Maybe I'm missing something. Let me know what you think.

File: frontend/src/metabase/visualizations/components/settings/ChartSettingOrderedColumns.jsx [hunk 3]
```diff
@@ -59,9 +59,11 @@ export default class ChartSettingOrderedColumns extends Component {
   };
 
   handleSortEnd = ({ oldIndex, newIndex }) => {
-    const fields = [...this.props.value];
-    fields.splice(newIndex, 0, fields.splice(oldIndex, 1)[0]);
-    this.props.onChange(fields);
+    const enabledColumns = [...this.props.value].filter(columnSetting =>
+      findColumnForColumnSetting(this.props.columns, columnSetting),
+    );
```
Comment (WiNloSt) [inline frontend/src/metabase/visualizations/components/settings/ChartSettingOrderedColumns.jsx:64]: I think I can safely use
```suggestion
    const enabledColumns = [...this.props.value].filter(columnSetting =>
      columnSetting.enabled
    );
```
silimiarly to what I've done in `frontend/src/metabase/visualizations/visualizations/Table.jsx`, but I use this since we have similar logic that yield a variable named `enabledColumns` in the `render` function.

File: frontend/test/metabase/scenarios/question/settings.cy.spec.js (hunk 2 above)
Comment (nemanjaglumac) [reply frontend/test/metabase/scenarios/question/settings.cy.spec.js:144]: Although Cypress has some default timeouts set in the config, **and** it comes with a promise of automatically (and intelligently) waiting, that didn't prove to be enough in our tests. Metabase has a set of complex XHRs firing basically after every little interaction. Some routes and/or actions have a predictable order of those requests, which we learned to use to our advantage. Btw, an anti-pattern in Cypress is to use the arbitrary wait (like `cy.wait(1000);`)

`POST /api/dataset` is one of the most commonly used ones. It's basically an ad-hoc query. If we wait for it, that means we've got the query results back. Sometimes (rarely) there is a slight delay from the time response comes back to the time browser renders the contents of that query. That's why you'll see `cy.findByTextEnsureVisible` helper in use sporadically.

> For me spy would make more sense when you want to:
>  - Simulate race condition
>  - Other complex states due to asynchronous actions

Ant that is exactly what's happening.
Cypress is infinitely faster than the real user UI interaction so we're running into race conditions. I agree it's muddying the waters by using the implementation details, but in this case that seems legit. The level of implementation details that I'd call unacceptable is when you assert on the state of some of the react components, for example. This is different, imo.

Wre're still asserting on what users see - just making sure results loaded first.

One more thing:
It's much easier to abstract code and create useful helper functions if we make sure that the request came back, than if we have to know what the contents of that response is for each scenario. For example `openOrdersTable` and `openProductsTable` are both waiting for `dataset` response (as is **any** visitQuestionAdHoc` helper). There is no way of knowing which string will be in the UI for each of those scenarios.

Sorry for the long response.
I hope it makes sense.

File: frontend/test/metabase/scenarios/question/settings.cy.spec.js [hunk 4]
```diff
@@ -147,6 +151,93 @@ describe("scenarios > question > settings", () => {
       }
     });
 
+    // https://github.com/metabase/metabase/pull/21338#pullrequestreview-928807257
+    it("should allow columns ordering after column removal and addition via sidebar (metabase#13455)", () => {
```
[the remark this answers is not in the document]
Comment (nemanjaglumac) [reply frontend/test/metabase/scenarios/question/settings.cy.spec.js:155]: I'll have to take a deeper look, but it seems that this test is basically the exact copy of the previous scenario with the extra step. In that case, let's expand the original test rather than adding another one.

https://docs.cypress.io/guides/references/best-practices#Creating-tiny-tests-with-a-single-assertion

File: frontend/test/metabase/scenarios/question/settings.cy.spec.js (hunk 2 above)
Comment (WiNloSt) [reply frontend/test/metabase/scenarios/question/settings.cy.spec.js:144]: @nemanjaglumac Thanks for the thorough writeup, I really appreciate it. You made your point, I also have to agree that some utils like `visitQuestionAdHoc` are really useful, and it'd be hard to only rely on the UI alone to achieve that.

File: frontend/test/metabase/scenarios/question/settings.cy.spec.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (WiNloSt) [reply frontend/test/metabase/scenarios/question/settings.cy.spec.js:155]: Yeah, that's exactly what it does. I'll combine the new test with the previous scenario. I didn't do that in the beginning because I'm afraid I'll make the test bloated. But if that's what Cypress is suggesting then I'd stick to it.

[the remark this answers is not in the document]
Comment (WiNloSt) [reply frontend/test/metabase/scenarios/question/settings.cy.spec.js]: Fixed

**Comment by WiNloSt** [comment]
@nemanjaglumac This is ready for review again 🙏

**Review by flamber — APPROVED** [review]
Seems to be working exactly as expected now.

File: frontend/test/metabase/scenarios/question/settings.cy.spec.js [hunk 5]
```diff
@@ -130,18 +130,48 @@ describe("scenarios > question > settings", () => {
...
 
+      // https://github.com/metabase/metabase/pull/21338#pullrequestreview-928807257
+
+      // Add "Address"
+      cy.findByText("Address")
+        .siblings(".Icon-add")
+        .click();
       /**
        * Helper functions related to THIS test only
        */
```
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/question/settings.cy.spec.js:141]: Nit: This should've been moved down, just above the helper functions, but please don't bother with it now. No point in waiting for the whole CI once again just because of that.

File: frontend/test/metabase/scenarios/question/settings.cy.spec.js [hunk 6]
```diff
@@ -130,18 +130,48 @@ describe("scenarios > question > settings", () => {
...
 
+      // The result automatically load when adding new fields
+      cy.wait("@dataset");
+
+      // Refresh @sidebarColumns as we added a new field
+      cy.findByText("Click and drag to change their order")
+        .parent()
+        .find(".cursor-grab")
+        .as("sidebarColumns"); // Store all columns in an array
```
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/question/settings.cy.spec.js:150]: This was a great catch! Simply getting the previous alias wouldn't have worked here.

File: frontend/test/metabase/scenarios/question/settings.cy.spec.js [hunk 7]
```diff
@@ -130,18 +130,48 @@ describe("scenarios > question > settings", () => {
...
+
+      findColumnAtIndex("User → Address", 16).as("user-address");
```
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/question/settings.cy.spec.js:152]: FYI, you can also use index `-1` for the last element, and `-2` for the second to last element if you don't care exactly how many elements are in the array.

File: frontend/test/metabase/scenarios/question/settings.cy.spec.js [hunk 8]
```diff
@@ -130,18 +130,48 @@ describe("scenarios > question > settings", () => {
...
+      // Add "Address"
+      cy.findByText("Address")
+        .siblings(".Icon-add")
+        .click();
       /**
        * Helper functions related to THIS test only
        */
 
+      // The result automatically load when adding new fields
+      cy.wait("@dataset");
```
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/question/settings.cy.spec.js:144]: @flamber I've noticed something strange in the Metabase at this point. It fires **two** `POST /api/dataset` xhrs. One seems to be related to the reordering (adding the new column), and another one is actually returning the query. Seems a bit wasteful - not sure if that was intended or a code smell that have to be looked at?

**Review by nemanjaglumac — APPROVED** [review]
Left some minor comments which don't affect the overall solution.
Great work :)

File: frontend/test/metabase/scenarios/question/settings.cy.spec.js (hunk 7 above)
Comment (WiNloSt) [reply frontend/test/metabase/scenarios/question/settings.cy.spec.js:152]: Deeyum! That's neat. I'll do that, as that tells the story better.

File: frontend/test/metabase/scenarios/question/settings.cy.spec.js (hunk 8 above)
Comment (flamber) [reply frontend/test/metabase/scenarios/question/settings.cy.spec.js:143]: Yeah, as @WiNloSt also mentioned (also mentioned in the issue), it runs the query and actually shows the change, but then it shows the Run-overlay, which causes double execution.
I simply don't know how our code is structured and if it can easily be prevented, but we can open another issue for that after this has been fixed.

Comment (WiNloSt) [reply frontend/test/metabase/scenarios/question/settings.cy.spec.js:143]: @flamber Just tested. It seems anytime you add a new field it calls `/api/dataset` twice. And when you add a new field it doesn't even show the run-overlay (the "Play button"), it seems that's automatic. I'll see if there's a quick win.

**Comment by WiNloSt** [comment]
FYI the latest change failed on `e2e-tests-question-ee` one time https://app.circleci.com/pipelines/github/metabase/metabase/32660/workflows/d5485db8-5d1c-4ce1-ad2b-2a2058a692fa/jobs/1674162/steps

**Comment by flamber** [comment]
Should probably have a FE dev check it for good measure. And should this just go in 43?

**Comment by WiNloSt** [comment]
@flamber my bad. I forgot to add fe devs as reviewer. And sounds good to not backport this. I trust your call.
